<launch>
  <master auto="start"/>
  <param name="/use_sim_time" value="false"/>
  <arg name="experiment_name" default="curve_manual2"/>
  <arg name="date" default="14-11-19"/>
  <arg name="open_rviz" default="false"/>
  <arg name="open_gazebo" default="false"/>
  <arg name="record_data" default="true" />

  <arg name="vl_force" default="true" />
  <arg name="rotational_force" default="true" />
  <arg name="inter_agent_force" default="true" />
  <arg name="motion_generator" default="false" />
  <!-- Define Paramerters here for robot formation-->
  <arg name="Kvl" default="100" />
  <arg name="Cvl" default="8" />
  <arg name="Kij" default="25" />
  <arg name="Cij" default="2" />
<!-- K angular should be low to let robot move in cartesion in time-->
  <arg name="K_ang" default="170" />
  <arg name="D_ang" default="0" />
  <arg name="team_size" value="2"/>
<!--
  <arg name="v_linear_threshold" value="0.2"/>
  <arg name="v_angular_threshold" value="0.2"/>
  <arg name="v_linear_cutoff" value="0.005"/>
  <arg name="v_angular_cutoff" value="0.05"/>
-->
  <arg name="robot_mass" value="40"/>
  <arg name="robot_radious" value="0.5"/>
  <arg name="robot_inertia" value="0.77"/>
  <arg name="delta_time" value="1"/> 
<!-- other robot map-->
<arg name="watkin_map" default="$(find multi_amr_nectec)/maps/watkin_lab.yaml"/>
<arg name="watkin_world" default="$(find multi_amr_nectec)/worlds/watkinslab.world"/>
 <arg name="map_empty" default="$(find multi_amr_nectec)/maps/map.yaml"/>
 <arg name="emptyworld" default="$(find turtlebot3_gazebo)/worlds/empty.world"/>
<arg name="neo_track1_world" default="$(find neo_simulation)/worlds/neo_track1.world"/>
<arg name="neo_track1_map" default="$(find neo_simulation)/maps/neo_track1.yaml"/>
<arg name="asphalt_plane" default="$(find multi_amr_nectec)/worlds/asphalt_plane.world"/>
<arg name="formation_test_world" default="$(find multi_amr_nectec)/worlds/formation_test.world"/>
<arg name="formation_test_map" default="$(find multi_amr_nectec)/maps/formation_test.yaml"/>
  <!-- Map server 
 <node pkg="map_server" type="map_server" name="map_server" args="$(arg formation_test_map)" respawn="false" >
    <param name="frame_id" value="/map" />
  </node>--> 
  <!-- start gazebo --> 

 <!-- RVIZ  --> 
  <group if="$(arg open_rviz)"> 
   <node pkg="rviz" type="rviz" name="rviz" required="true"
         args="-d $(find multi_amr_nectec)/rviz/mocap_rviz_nin.rviz"/>
  </group>

<!-- AMR 0 -->
<group ns="amr_0">
<node pkg="tf" type="static_transform_publisher" name="optitrack_remap_amr_0" args="0 0 0 0 0 0 /mocap_amr_0 amr_0/base_footprint 10"/>

</group>

<!-- AMR 1 -->
<group ns="amr_1">
<node pkg="tf" type="static_transform_publisher" name="optitrack_remap_amr_1" args="0 0 0 0 0 0 /mocap_amr_1 amr_1/base_footprint 10"/>

</group>

 <group ns="vl_robot">
 <!--VL localization -->
  <!--room edge  x -0.8 y -1.75 -->
  <!--room centre  x -0.8 y -0.3 -->
 <node pkg="multi_amr_nectec" type="fake_odom_vl" name="fake_odom_vl" output="screen">
  <param name ="default_x" value = "-0.8" />
  <param name ="default_y" value = "-1.75" />
  <param name ="default_theta" value = "1.57079" />
</node>
 <!--transform virtual leader odom to world-->
 <node pkg="tf" type="static_transform_publisher" name="vl_odom_world_broadcaster" args="0 0 0 0 0 0 /world /vl_robot/odom 10" />
<!--Get odom from virtual leader robot  -->
<node pkg="multi_amr_nectec" type="get_odom_robot" name="get_odom_robot" output="screen" />
 <!--Get angle from virtual leader robot  -->
<node pkg="multi_amr_nectec" type="tf_quat2euler" name="tf_quat2euler" output="screen" />
</group>



<!--Load parameter     -->
<rosparam command="load" file="$(find multi_amr_nectec)/config/formation_shape/2_amr_linear.yaml" />


<!-- VL force     -->
<node pkg="multi_amr_nectec" type="vl_force" name="vl_force" output="screen" if="$(arg vl_force)">
  <param name ="team_size" value = "$(arg team_size)" />
  <param name ="Kvl" value = "$(arg Kvl)" />
  <param name ="Cvl" value = "$(arg Cvl)" />
  <param name ="delta_time" value = "$(arg delta_time)" />
</node>
<!--Rotational force   --> 
<node pkg="multi_amr_nectec" type="orientation_control_exper" name="orientation_control_exper" output="screen" if="$(arg rotational_force)" >
  <param name ="team_size" value = "$(arg team_size)" />
  <param name ="K_ang" value = "$(arg K_ang)" />
    <param name ="D_ang" value = "$(arg D_ang)" />
  <param name ="delta_time" value = "$(arg delta_time)" />
</node>
<!-- Inter agent force     -->
<node pkg="multi_amr_nectec" type="inter_agent_force" name="inter_agent_force" output="screen" if="$(arg inter_agent_force)">
  <param name ="team_size" value = "$(arg team_size)" />
  <param name ="Kij" value = "$(arg Kij)" />
  <param name ="Cij" value = "$(arg Cij)" />
  <param name ="delta_time" value = "$(arg delta_time)" />
</node>

 
<!--motion generator -->
<node pkg="multi_amr_nectec" type="motion_generator" name="motion_generator" output="screen" if="$(arg motion_generator)" >
 <param name ="team_size" value = "$(arg team_size)" />
<param name ="v_linear_threshold" value = "$(arg v_linear_threshold)" />
<param name ="v_angular_threshold" value = "$(arg v_angular_threshold)" />
<param name ="v_linear_cutoff" value = "$(arg v_linear_cutoff)" />
<param name ="v_angular_cutoff" value = "$(arg v_angular_cutoff)" />
<param name ="robot_mass" value = "$(arg robot_mass)" />
<param name ="robot_radious" value = "$(arg robot_radious)" />
<param name ="robot_inertia" value = "$(arg robot_inertia)" />
<param name ="delta_time" value = "1" />
</node>
<!--Publish X Y Theta of follower robot-->
<node pkg="multi_amr_nectec" type="pose2d_real_robot" name="pose2d_real_robot" output="screen" />
<!--ROSBAG Record -->
 <node pkg="rosbag" type="record" name="rosbag_record_data" output="screen"
       args="record -O /home/kanin/ROSbag/$(arg date)_$(arg experiment_name)_Kvl=$(arg Kvl)_Cvl=$(arg Cvl)-Kij=$(arg Kij)-Cij=$(arg Cij)-K_ang=$(arg K_ang).bag 
       /vrpn_client_node/mocap_amr_0/pose amr_0/pose2d /amr_0/cmd_vel  /vrpn_client_node/mocap_amr_1/pose amr_1/pose2d /amr_1/cmd_vel /vl_robot/pose2d 
       /vrpn_client_node/object_load/pose /vl_robot/cmd_vel /F_rot /F_vl_x /F_vl_y /F_ij_x /F_ij_y "
       if="$(arg record_data)" /> 

<!-- Linear trajectory of VL  
<node pkg="multi_amr_nectec" type="vl_trajectory_linear" name="vl_trajectory_linear" output="screen" />
--> 
<!--trajectory of VL S curve
<node pkg="multi_amr_nectec" type="vl_trajectory_s_curve" name="vl_trajectory_s_curve" output="screen" />
--> 
 <!--trajectory of VL circle 
<node pkg="multi_amr_nectec" type="vl_trajectory_circle" name="vl_trajectory_circle" output="screen" />
 -->
<!-- Spawn load  
 <include file="$(find multi_amr_nectec)/launch/spawn_load.launch" >
   </include>
-->
</launch>